src/startup/ModeCLIParser.h

```diff
@@ -0,0 +1,128 @@
+        return make_shared<ValueArg<uint>>(_short, _long, _desc, req, _default, "uint", *parser);
+    }
+
+    static _UInt64ValueArg
+    _makeUInt64ValueArg(const string& _short, const string& _long, const string& _desc, bool req, int _default, CmdLine *parser) {
+        return make_shared<ValueArg<int64_t>>(_short, _long, _desc, req, _default, "int64_t", *parser);
+    }
+
+    static _SwitchArg
+    _makeSwitchArg(const string& _short, const string& _long, const string& _desc, CmdLine *parser, bool _default = false) {
+        return make_shared<SwitchArg>(_short, _long, _desc, *parser, _default);
+    }
+
+public:
+
+    /**
+     * Needs to be overriden by specialized mode parsers and is used to parse the application arguments.
+     * Note, that subclasses may use derived classes of Runner as a return type (C++ allows this). However, e.g. CLion
+     * will not recognize parse as used!
+     */
+    virtual Runner *parse(int argc, const char **argv) = 0;
+
+    shared_ptr<CmdLine> createCommandLineParser() {
+        return make_shared<CmdLine>(
+                "FastqInDex - A tool to index compressed FASTQ (or text files) and to allow random data extraction from them.",
+                '=', "0.10b", false);
+    }
+
+    _IntValueArg createVerbosityArg(CmdLine *cmdLineParser) const;
+
+    _SwitchArg createDebugSwitchArg(CmdLine *cmdLineParser) const;
+
+    _StringValueArg createIndexFileArg(CmdLine *cmdLineParser) const;
+
+    _StringValueArg createFastqFileArg(CmdLine *cmdLineParser) const;
+
+    _StringValueArg createS3CredentialsFileArg(CmdLine *cmdLineParser) const;
+
+    _StringValueArg createS3ConfigFileArg(CmdLine *cmdLineParser) const;
+
+    _StringValueArg createS3ConfigFileSectionArg(CmdLine *cmdLineParser) const;
+
+    _SwitchArg createForceOverwriteSwitchArg(CmdLine *cmdLineParser) const;
+
+    tuple<shared_ptr<UnlabeledValueArg<string>>, shared_ptr<ValuesConstraint<string>>>
+    createAllowedModeArg(const string &mode, CmdLine *cmdLineParser) const;
+    
+    static bool isS3Path(const string& str) {
+        string_view _str = str;
+        string prefix = "s3://";
+        return _str.size() > prefix.size() && _str.substr(0, prefix.size()) == prefix;
+    }
+
+    static string resolveIndexFileName(const string &file, const shared_ptr<Source> &fastq) {
+        // S3 and Path based! Stream is not a valid input here, as it has no name. In that case, the original value will
+        // be returned.
+        if (file.empty() && fastq->isFile()) {
+            return fastq->toString() + ".fqi";
+        }
+        return file;
+    }
+
+    /**
+     * Processes several FASTQ related parameters and create a FASTQ PathSource from the input.
+     */
+    static shared_ptr<Source> processFastqFile(const string &fastqFileArg, const S3ServiceOptions &s3ServiceOptions);
+
+    static shared_ptr<Source> processIndexFileSource(const string &indexFile,
```

Comment:
As layout Kevlin Henney suggests e.g.
```c++
static shared_ptr<Source> processIndexFileSource(
   const string &indexFile,
   const shared_ptr<Source> &fastqSource,
   const S3ServiceOptions &s3ServiceOptions);
```

i.e. with all parameters in new lines (more tweaks not excluded), because then the interesting information is not at the remote right and the eye does not have to move so far right.